Source/WebCore/Modules/WebGPU/GPUBuffer.cpp

```diff
@@ -42,15 +56,48 @@ void GPUBuffer::setLabel(String&& label)
 
 void GPUBuffer::mapAsync(GPUMapModeFlags mode, std::optional<GPUSize64> offset, std::optional<GPUSize64> size, MapAsyncPromise&& promise)
 {
-    m_backing->mapAsync(convertMapModeFlagsToBacking(mode), offset.value_or(0), size, [promise = WTFMove(promise)] () mutable {
+    if (!m_bufferSize || (size.has_value() && !size.value())) {
         promise.resolve(nullptr);
+        return;
+    }
+
+    if (m_pendingMapPromise) {
+        promise.reject(Exception { OperationError });
+        return;
+    }
+
+    if (m_mapState == GPUBufferMapState::Unmapped)
+        m_mapState = GPUBufferMapState::Pending;
+
+    m_pendingMapPromise = promise;
+    m_backing->mapAsync(convertMapModeFlagsToBacking(mode), offset.value_or(0), size, [promise = WTFMove(promise), strongThis = Ref { *this }] (bool success) mutable {
```

Comment:
We normally use `protectedThis` in C++ code (and `strongSelf` in ObjC++ code).

Reply:
Thanks I'll change that